### DISCUSSION TITLE
Groundhog year

### DISCUSSION THREAD

**Discussion by TrueBrain** [OpenTTD/OpenTTD#8384] [Features]
A while back we got the feature request to add "groundhog year", a year which repeats over and over and over and over.

A [Pull Request](https://github.com/OpenTTD/OpenTTD/pull/7938) was made, but the intention were not described. This means different people have different expectation about such feature.

Additionally, code-wise, a few issues arise. For example, inflation normally runs for only 170 years. With groundhog year, this can be infinite long. Is that a problem? Is that intended? Nobody knows, as it wasn't really discussed.

So, here we have a discussion thread! We would love people to test out groundhog year (see PR above for code), and report their findings. What was acting up? What was completely weird? And all that in many different scenarios. Different settings. Obscure NewGRFs. Etc.

Please avoid: WE SHOULD DO THIS, statements. We rather hear what you like and dislike, the problems you found, and the things you think should be fixed. No demands. Open discussion :)

Thank you for participating!

**Comment by andythenorth** [comment]
I have NFI how this discussions will work, but let's try it!  Forward!

Things I can think of that might break follow, one heading per comment.  No particular order.

See also: #8397 (daylength) as groundhog year is an alternative route to some goals that daylength tries to meet.

**Comment by andythenorth** [comment]
### Inflation
Inflation! My opinions about inflation are too negative to be useful here.  So I ask: if inflation is 'on' in a groundhog year, what effects, if any are there?

**Reply by TrueBrain** [reply]
The code defined that inflation runs for 170 years (originally, the game ran from 1920 to 2090); after this it stops. This is calculated from with `current-year minus start-year-of-the-map`.

So if a groundhog year happens before the 170 years run out, inflation keeps happening every month till the inflation limit is reached.

**Reply by 2TallTyler** [reply]
I think this could easily solved by adding a check to [AddInflation()](https://github.com/OpenTTD/OpenTTD/blob/master/src/economy.cpp#L726) which skips increasing inflation, in the same way it is skipped after 170 years or the inflation limit is reached:
`if (_cur_year == _settings_game.game_creation.loop_year) return true;`

**Comment by andythenorth** [comment]
### Vehicle introduction
I often play a style where I reset the year with the date cheat.  I believe this can cause recently introduced vehicles to disappear again from the buy menu.  Does that happen with groundhog year?

**Reply by andythenorth** [reply]
I add, when I play games, I am constantly using reload_newgrfs which re-randomises intro dates, so my observed behaviour in previous games is not much use on this item.

**Comment by andythenorth** [comment]
### Vehicle exclusive trial period offers
Does it work?

**Comment by andythenorth** [comment]
### Vehicle reliability change and retirement (removal from buy menu)
Does groundhog year affect reliability change, reliability decay period, and eventual retirement (removal from buy menu).  These are related items in the spec so I include them together under this heading.

**Reply by frosch123** [reply]
I see these options here:
* Freeze reliability/retirement in the groundhog year: newly introduced vehicles remain unreliable unless you advance the setting manually.
* Continue increasing reliability of new vehicles, but freeze decrease/retirement: old vehicles remain unreliable and remain in the purchase list.
* Only freeze engines that are neither new nor about to retire. This is probably so complicated, that it raises follow-up questions.

**Comment by andythenorth** [comment]
### Vanilla industry behaviour
Vanilla industries *should* produce and close just fine, based on what I know of the industry code.  Do they though?

What about e.g. oil wells that close?  Do they fail to close if they are marked for closure, and then the year is reset?

What about industries with intro dates?  E.g. Oil Rigs?  If the groundhog year co-incides with intro date, do they become inconsistently available?

**Comment by andythenorth** [comment]
### Newgrf industry behaviour

Newgrf industries like ECS, FIRS and PBI may check the dates widely, example to determine availability, closure, production behaviour, appearance of graphics, etc. Their behaviour is very subject to newgrf author choices. 

Is this even in scope for testing with groundhog year?

**Comment by andythenorth** [comment]
### Newgrf vehicles behaviour

Similar to newgrf industries, newgrf vehicles can check the date for many things.  Does that have problems?

**Comment by andythenorth** [comment]
### Newgrf xxx
Rinse and repeat, most newgrf features can check the game date as a var in a varact2 chain.  So are we putting objects, station, houses, railtypes, roadtypes etc in scope for testing with this feature?

**Comment by andythenorth** [comment]
### Subsidies
How does subsidy issuing / awarding / expiring work with groundhog year?

**Comment by andythenorth** [comment]
### Colour news vs. monochrome news
There is a date setting for this, does it flip-flop with groundhog year?  Does it matter?

**Comment by andythenorth** [comment]
### Semaphore signal change date
There is a date setting for this, does it flip-flop with groundhog year?  Does it matter?

**Comment by andythenorth** [comment]
### Game ending year
What happens if game ending year is also ground year (e.g. 2050)?  Does the game end screen show repeatedly?

**Reply by TrueBrain** [reply]
Or worse: does it keep adding your name on the highscore board?

**Comment by andythenorth** [comment]
### GameScripts
GS can use date extensively, for example goal scripts like NoCarGoal, Silicon Valley and Busy Bee all use date for goal achievements.  How do they handle groundhog year?  What breaks?  Does their internal state become corrupt?

**Comment by andythenorth** [comment]
### Finances window
What does the finance window do in a groundhog year?  How does it calculate 'previous years', if at all?

**Comment by andythenorth** [comment]
### Loan interest / calculations
Are there any effects on the loan interest calculation? I suspect there shouldn't be, it should just be monthly, but may need checked.

**Comment by andythenorth** [comment]
### Game charts
How do the charts handle groundhog year.  E.g. Operating profit graph, income graph etc.

**Comment by andythenorth** [comment]
### Town ratings (including effect of bribery)
I believe town ratings don't use a specific date, but are a monthly calculation, and should be fine with groundhog year.  Is this correct?

**Comment by andythenorth** [comment]
### This list is not complete?
This is what I could think of.  What did I miss?

**Comment by KeepinItRail** [comment]
What does groundhog year do when you ‘Fund new buildings’, ‘Launch an ad campaign’, etc from a Town window, all of which normally last for a certain time period?

**Reply by michicc** [reply]
They last the same time period. I don't think anything else would make much sense.
